### PR TITLE
Adds support for running native binaries without JS intermediaries

### LINKED ISSUE #882: [Bug] SyntaxError when running a "bin" file that isn't authored in JS
- [ ] I'd be willing to implement a fix

**Describe the bug**

Yarn 2.0 fails with a `SyntaxError` when running a `bin` file that isn't authored in JS. For example, the bin file may be an executable file with a shebang `#!/usr/bin/env sh`.

Example `bin` file:

```sh
#!/usr/bin/env sh
echo 'Hello world!'
```

**To Reproduce**

See sample package ([`@cameronhunter/berry-bin-bug`](https://github.com/cameronhunter/berry-bin-bug)) that defines a `bin` that is an [executable file with a shebang](https://github.com/cameronhunter/berry-bin-bug/blob/master/bin/berry-bin-bug).

```js repro
const { promises: { writeFile, chmod } } = require('fs');

await writeFile('./hello-world', '#!/usr/bin/env sh\necho "Hello world!"');
await chmod('./hello-world', 0o765);

await packageJsonAndInstall({
  bin: { 'hello-world': './hello-world' }
});

await expect(yarn('hello-world')).resolves.toContain('Hello world!');
```

Example of failure:

```sh
$ yarn berry-bin-bug
/Users/chunter/workspace/github/cameronhunter/berry-bin-bug/bin/berry-bin-bug:3
echo "Hello world!"
     ^^^^^^^^^^^^^^

SyntaxError: Unexpected string
    at Module._compile (internal/modules/cjs/loader.js:895:18)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:995:10)
    at Module.load (internal/modules/cjs/loader.js:815:32)
    at Function.module_1.Module._load (/Users/chunter/workspace/github/cameronhunter/berry-bin-bug/.pnp.js:13519:14)
    at Function.Module.runMain (internal/modules/cjs/loader.js:1047:10)
    at internal/main/run_main_module.js:17:11
```

**Environment if relevant (please complete the following information):**

 - OS: MacOS Mojave (`10.14.5`)
 - Node version: `12.14.0`
 - Yarn version: `2.0.0-rc.28`

### PR DESCRIPTION
**Pull request by arcanis**
**What's the problem this PR addresses?**

Yarn currently cannot run native binaries without going through a JS jumper script. Other package managers don't have this restriction, and it makes the `yarn run` overhead worse on some scripts for little reasons - or doesn't work at all when packages don't use jumper scripts.

Fixes #882

**How did you fix it?**

Two mechanisms are used:

- First we check for the binary extension
- Then we check its magic number

If one of the two match a predetermined list, the binary is spawned without going through Node. This ensures that this logic is called only when the binary is truly a native binary, and will not affect the behaviour of other existing scripts.

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/plugin-pnp/sources/jsInstallUtils.ts [hunk 1]
```diff
@@ -58,6 +58,7 @@ export function extractBuildScripts(pkg: Package, requirements: ExtractBuildScri
 }
 
 const FORCED_EXTRACT_FILETYPES = new Set([
+  `.bin`,
```
Comment (merceyz) [inline packages/plugin-pnp/sources/jsInstallUtils.ts:61]: Could you add a comment for this extension?

**Review by merceyz — CHANGES_REQUESTED** [review]
Presumably we'll need to do this for the following locations as well:
https://github.com/yarnpkg/berry/blob/ca475757fe85cdc2c4b4441f6d65d38fdab230ad/packages/yarnpkg-core/sources/scriptUtils.ts#L491-L495
https://github.com/yarnpkg/berry/blob/ca475757fe85cdc2c4b4441f6d65d38fdab230ad/packages/yarnpkg-core/sources/scriptUtils.ts#L544-L548
https://github.com/yarnpkg/berry/blob/ca475757fe85cdc2c4b4441f6d65d38fdab230ad/packages/yarnpkg-core/sources/scriptUtils.ts#L741-L745

**Comment by merceyz** [comment]
There are some issues that this probably fixes (for example https://github.com/yarnpkg/berry/issues/882), could you check and link to them so they're closed?

**Comment by jmatsushita** [comment]
I've tried `yarn set version 4.0.0-rc.47` with `expo@9.0.0-beta.0` in a monorepo but it doesn't fix ti.
```
> yarn workspace client expo info                                                                                                       .../node_modules/expo/bin/cli:3
# Forward all arguments to the local CLI (@expo/cli).
^

SyntaxError: Invalid or unexpected token
    at internalCompileFunction (node:internal/vm:73:18)
    at wrapSafe (node:internal/modules/cjs/loader:1176:20)
    at Module._compile (node:internal/modules/cjs/loader:1218:27)
    at Module._extensions..js (node:internal/modules/cjs/loader:1308:10)
    at Module.load (node:internal/modules/cjs/loader:1117:32)
    at Module._load (node:internal/modules/cjs/loader:958:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:23:47

Node.js v18.16.0
```

**Comment by arcanis** [comment]
This fix is only about *native binaries*. From your logs, expo is a shellscripts, which aren't allowed in `bin` at the moment for the reasons outlined in https://github.com/yarnpkg/berry/issues/882#issuecomment-1599478009.

**Comment by MorisR** [comment]
Will the changes be also added to yarn 3? 
we're still stuck on node v16 and yarn 4 requires node v18

**Comment by merceyz** [comment]
Yes, I've backported it now so it will be in the next release.

**Pull request merged**